odoo/tools/lru.py

```diff
@@ -31,29 +35,62 @@ def get(self, obj, val=None):
 
     @locked
     def __getitem__(self, obj):
-        a = self.d[obj]
+        a = self.d[obj][0]
         self.d.move_to_end(obj, last=False)
         return a
 
     @locked
     def __setitem__(self, obj, val):
-        self.d[obj] = val
+        size = get_memory_footprint(obj, val) if self.mode == 'footprint' else 1
+        self.d[obj] = (val, size)
         self.d.move_to_end(obj, last=False)
-        while len(self.d) > self.count:
-            self.d.popitem(last=True)
+        self.current_size += size
+        while self.current_size > self.maximum_size:
+            self.current_size -= self.d.popitem(last=True)[1][1]
 
     @locked
     def __delitem__(self, obj):
-        del self.d[obj]
+        self.pop(obj)
 
     @locked
     def __len__(self):
         return len(self.d)
 
     @locked
-    def pop(self,key):
-        return self.d.pop(key)
+    def pop(self, key):
+        value, size = self.d.pop(key)
+        self.current_size += size
+        return value
 
     @locked
     def clear(self):
         self.d.clear()
+        self.current_size = 0
+
+
+def get_memory_footprint(*args):
+    seen_ids = set()
+    total_size = 0
+    objects = collections.deque(args)
+
+    while objects:
+        cur_obj = objects.popleft()
+        if id(cur_obj) in seen_ids or callable(cur_obj):
+            continue
+
+        seen_ids.add(id(cur_obj))
+        total_size += sys.getsizeof(cur_obj)
```

Comment:
There are some edge cases to handle. This is the official example: https://code.activestate.com/recipes/577504-compute-memory-footprint-of-an-object-and-its-cont/